### ISSUE TITLE
Random crashes with 0-bytes gzip files

### ISSUE THREAD

**Issue by fjossandon** [marcelm/cutadapt#514]
Hello,
`This is cutadapt 3.2 with Python 3.8.7`, installed with PIP.
I'm using Cutadapt to remove primers from the Gzipped Fastq files generated by an older version of Bcl2fastq which sometimes generates totally empty files (0 bytes). I think that the reason for this, is because that Bcl2fastq version creates empty files with the filename including ".gz" at the beginning and starts filling them during the execution (newer versions don't do this anymore, but for historical reasons I cannot update it). So if it cannot find the barcode pair in the SampleSheet, the resulting files are like this:
```
96894_S269_L001_R1_001.fastq.gz   0 bytes
96894_S269_L001_R2_001.fastq.gz   0 bytes
```
You can replicate them using "touch filename.gz". This is different than generating an empty file first with "touch filename", and then applying gzip on it, in which case the final file will have a few bytes of metadata from gzip. Using a zcat on the 0 bytes files always results in a gzip error:
```
$ zcat 96894_S269_L001_R1_001.fastq.gz

gzip: 96894_S269_L001_R1_001.fastq.gz: unexpected end of file
```

The thing is that I parallelize the processing of the Fastq files from a sequencing run in which many samples have 0 bytes. In most cases Cutadapt process them gracefully and give a "**No reads processed!**" message (which is what I expected), but a few randomly crash. For example with this command:
* cutadapt -a 'AACTTTYARCAAYGGATCTC;max_error_rate=0.1;min_overlap=20' -A 'TGATCCYTCCGCAGGT;max_error_rate=0.5;min_overlap=16' --pair-adapters --pair-filter any --cores 2 --output /pipeline/output/96894_R1.fastq --paired-output /pipeline/output/96894_R2.fastq /pipeline/input/96894_S269_L001_R1_001.fastq.gz /pipeline/input/96894_S269_L001_R2_001.fastq.gz
```
Traceback (most recent call last):
  File "/usr/local/bin/cutadapt", line 8, in <module>
    sys.exit(main_cli())
  File "/usr/local/lib/python3.8/site-packages/cutadapt/__main__.py", line 845, in main_cli
    main(sys.argv[1:])
  File "/usr/local/lib/python3.8/site-packages/cutadapt/__main__.py", line 912, in main
    stats = r.run()
  File "/usr/local/lib/python3.8/site-packages/cutadapt/pipeline.py", line 824, in run
    raise e
OSError: gzip: /pipeline/input/96894_S269_L001_R1_001.fastq.gz: unexpected end of file (exit code 1)
```
The non-reproducible part is that if I take these files and execute Cutadapt again, it usually works and prints the "**No reads processed!**", and when retrying the whole folder, it usually crashes in other random cases.

So in summary, there is something that I have not been able to determine that makes Cutadapt crash sometimes when dealing with 0 bytes ".gz" files. A workaround I will be implementing on my side is to check the file sizes first, and avoid calling Cutadapt if they are 0 bytes. But initial tests made me thought that Cutadapt was robust and could handle these cases well, so  crashes were a surprise. Maybe you can pinpoint the root problem or make the same workaround of pre-checking the file sizes to avoid the crash.

Best regards,

**Comment by marcelm** [comment]
Thanks! This is probably a bug in xopen, which Cutadapt uses. That bug may have been fixed recently, but I’ll have to perhaps adjust some things in Cutadapt also.

In your case, Cutadapt starts up a gzip process to decompress the input file. The bug is that errors that gzip produces are often not caught. But "fixing" this would imply that you always get an error message, which is also not what you want.

For other reasons, I think the correct fix is to no longer use an external gzip process to decompress the input (that was done for speed reasons at some point, but is no longer relevant).

A quick workaround: Pretend that you have no `gzip` installed by using an empty `$PATH` when you run `cutadapt`:
```
PATH= cutadapt ...
```
(Note the space after `=` in case you’re not familiar with this syntax.) This will force Cutadapt to use Python’s built-in decompression routines.

I’ll try to fix this later.

**Comment by fjossandon** [comment]
> In your case, Cutadapt starts up a gzip process to decompress the input file. The bug is that errors that gzip produces are often not caught. But "fixing" this would imply that you always get an error message, which is also not what you want.

Yes, my ideal scenario is that those 0 bytes files finish normally and trigger the "No reads processed!" message. =)
Thanks for the tip!

**Comment by marcelm** [comment]
I have fixed this in c20cc5b4bafe816d67ae777eba0ad3ac80d9a967: Cutadapt will no longer use an external process for decompressing input files. I think I’ll make a new release soon.
